ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala

```diff
@@ -1493,6 +1501,243 @@ object Serializers {
           optimizerHints)
     }
 
+    private def jlClassMethodsHack16(methods: List[MethodDef]): List[MethodDef] = {
+      for (method <- methods) yield {
+        implicit val pos = method.pos
+
+        val methodName = method.methodName
+        val methodSimpleNameString = methodName.simpleName.nameString
+
+        val thisJLClass = This()(ClassType(ClassClass, nullable = false))
+
+        if (methodName.isConstructor) {
+          val newName = MethodIdent(NoArgConstructorName)(method.name.pos)
+          val newBody = ApplyStatically(ApplyFlags.empty.withConstructor(true),
+              thisJLClass, ObjectClass, newName, Nil)(NoType)
+          MethodDef(method.flags, newName, method.originalName,
+              Nil, NoType, Some(newBody))(
+              method.optimizerHints, method.version)
+        } else if (methodSimpleNameString == "newArrayOfThisClass") {
+          val newName = MethodIdent(MethodName(methodName.simpleName,
+              List(IntRef), ClassRef(ObjectClass)))(method.name.pos)
+          val lengthParam = ParamDef(LocalIdent(LocalName("length")),
+              NoOriginalName, IntType, mutable = false)
+          val newBody = BinaryOp(BinaryOp.Class_newArray, thisJLClass, lengthParam.ref)
+          MethodDef(method.flags, newName, method.originalName,
+              List(lengthParam), AnyType, Some(newBody))(
+              method.optimizerHints.withInline(true), method.version)
```

Comment:
Would you be very opposed to expose `CheckNull` as a `UnaryOp`? Then this (and the strange private method in `Class`) would go away.

Reply:
No I think it definitely makes sense. There's a bit of open design questions for that, though: 

* If we make it a `UnaryOp`, we cannot determine the `tpe` of a `UnaryOp`'s `op` alone anymore. It now depends on the type of the argument. Should this be grounds to make it its own node instead? Not sure. (Note that `Clone` will have the same issue if we want to turn it into a `UnaryOp`.)
* Instead of a new kind of operation, should it be merged into the behavior of `AsInstanceOf`? If the target type is not nullable (but also not a primitive because that's an unboxing operation) and the argument's type is nullable, then the behavior would include a null check. That has the useful side effect that we can express an `AsInstanceOf` from non-nullable to non-nullable as well, basically for free. However it's a more involved change. 

Do you have any opinion about these alternatives?

Reply:
> Should this be grounds to make it its own node instead?

I'm not sure. It feels like that change would still be extremely local. All usage sites are going to look at `tpe` anyways if they need the type.

> Instead of a new kind of operation, should it be merged into the behavior of AsInstanceOf?

I think conceptually this would be cleaner, yes. I have been fiddling with this and a couple of challenges surface already:

- We need more optimization in the emitter stage: Checking for null and for a type are very different at the emitter level. So always doing both is wasteful.
- If we naively replace `CheckNotNull(expr)` with `AsInstanceOf(expr, expr.tpe.toNonNullable)` we get undefined references (at the GCC level) for `$as_jl_Long` and `$as_jl_Throwable`. The former is probably the fault of the optimizer (runtime long substitution is the last thing IIUC that prevents us from running IR checking post optimizer). The latter is probably the fault of the emitter (JS throwable replacement).
- Side-effect checking for `AsInstanceOf` under semantics becomes more complicated: We now depend on two semantics (asInstanceOf and NPE). Again, if we do the checking naively (require both semantics to be unchecked), we'll almost certainly regress in code size.

So all-in-all, I'm not sure we're better of with the `AsInstanceOf` option (but maybe more testing is required). I  think what we should definitely do is allow non-nullable types in `AsInstanceOf` if the target expr is already non-nullable (but IIUC that is a trivial change).

Reply:
I agree with the analysis. It seems merging with `AsInstanceOf` causes more issues than it solves.

I made the unary op version in #5037.